File: src/main/java/com/machfour/macros/objects/QtyUnit.java
package com.machfour.macros.objects;

import com.machfour.macros.core.*;
import org.jetbrains.annotations.NotNull;

import static com.machfour.macros.core.Schema.QtyUnitTable.*;

/*
 * Units for measuring quantities of food (only). Not for nutrition measurements.
 */
public class QtyUnit extends MacrosEntity<QtyUnit> implements Measurement<QtyUnit>, Unit {

    public static Factory<QtyUnit> factory() {
        return QtyUnit::new;
    }

    public static Table<QtyUnit> table() {
        return Schema.QtyUnitTable.instance();
    }

    @Override
    public String toString() {
        return String.format("%s (%s)", name(), abbr());
    }


    private QtyUnit(ColumnData<QtyUnit> data, ObjectSource objectSource) {
        super(data, objectSource);
    }


    @Override
    public Table<QtyUnit> getTable() {
        return table();
    }

    @Override
    public Factory<QtyUnit> getFactory() {
        return factory();
    }

    @NotNull
    @Override
    public String name() {
        return getData(NAME);
    }

    @NotNull
    @Override
    public String abbr() {
        return getData(ABBREVIATION);
    }

    @NotNull
    public Double metricEquivalent() {
        return getData(METRIC_EQUIVALENT);
    }

    @NotNull
    public Boolean isVolumeUnit() {
        return getData(IS_VOLUME_UNIT);
    }

    // Measurement functions
    @Override
    public double unitMultiplier() {
        return 1;
    }

    @Override
    public QtyUnit baseUnit() {
        return this;
    }

    @Override
    public boolean isVolumeMeasurement() {
        return isVolumeUnit();
    }

    @Override
    public boolean isServing() {
        return false;
    }
}

File: src/main/kotlin/com/machfour/macros/objects/QtyUnit.kt
package com.machfour.macros.objects

import com.machfour.macros.core.*
import com.machfour.macros.core.Schema.QtyUnitTable.Companion.ABBREVIATION
import com.machfour.macros.core.Schema.QtyUnitTable.Companion.IS_VOLUME_UNIT
import com.machfour.macros.core.Schema.QtyUnitTable.Companion.METRIC_EQUIVALENT
import com.machfour.macros.core.Schema.QtyUnitTable.Companion.NAME

/*
 * Units for measuring quantities of food (only). Not for nutrition measurements.
 */
class QtyUnit private constructor(data: ColumnData<QtyUnit>, objectSource: ObjectSource)
    : MacrosEntityImpl<QtyUnit>(data, objectSource), PortionMeasurement, Unit {
    companion object {
        // factory before table
        val factory : Factory<QtyUnit> = Factory { dataMap, objectSource -> QtyUnit(dataMap, objectSource) }

        val table: Table<QtyUnit>
            get() = Schema.QtyUnitTable.instance
    }

    override val factory: Factory<QtyUnit>
        get() = Companion.factory

    override val table: Table<QtyUnit>
        get() = Companion.table

    val isVolumeUnit: Boolean
        get() = getData(IS_VOLUME_UNIT)!!

    override fun toString(): String {
        return "$name (${abbr})"
    }

    override val name: String
        get() = getData(NAME)!!

    override val abbr: String
        get() = getData(ABBREVIATION)!!

    fun metricEquivalent(): Double {
        return getData(METRIC_EQUIVALENT)!!
    }

    // Measurement interface
    override val unitMultiplier = 1.0
    override val baseUnit = this
    override val isVolumeMeasurement = isVolumeUnit

}
